### ISSUE TITLE
Loss Implementation

### ISSUE THREAD

**Issue by toshi2k2** [stefanopini/simple-HRNet#26]
I was playing around with your code and noticed that the loss and backprop is doing something strange. I didn't change anything in your code - ran validation before training the (trained) model for 1 epoch. That 1 epoch of training seemed to drag the results to almost 0. Any idea why that is happening?

`
Loading train and validation datasets...
loading annotations into memory...
Done (t=8.38s)
creating index...
index created!
100%|███████████████████████████████████████████████████████████████████████████| 118287/118287 [00:42<00:00, 2776.12it/s]

COCO dataset loaded!
loading annotations into memory...
Done (t=0.37s)
creating index...
index created!
100%|███████████████████████████████████████████████████████████████████████████████| 5000/5000 [00:02<00:00, 2203.10it/s]

COCO dataset loaded!
cuda:0

Training started @ 2019-10-13 14:48:47

Epoch 1 of 200 @ 2019-10-13 14:48:47
Validating: 100%|███████████████████████████████████████████████████████████████████████| 397/397 [12:05<00:00,  1.26s/it]

Val AP/AR
WARNING:root:json-tricks: numpy scalar serialization is experimental and may work differently in future versions
Loading and preparing results...
DONE (t=0.70s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *keypoints*
DONE (t=7.33s).
Accumulating evaluation results...
DONE (t=0.13s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.772
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.936
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.847
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.739
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.828
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.800
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.945
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.864
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.762
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.859
Training: 100%|█████████████████████████████████████████████████████████████████████| 9363/9363 [8:32:08<00:00,  2.00s/it]

Train AP/AR
Loading and preparing results...
DONE (t=17.94s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *keypoints*
DONE (t=136.31s).
Accumulating evaluation results...
DONE (t=3.30s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.011
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.034
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.004
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.011
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.011
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.037
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.105
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.020
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.034
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.040
Validating: 100%|███████████████████████████████████████████████████████████████████████| 397/397 [08:43<00:00,  1.27s/it]

Val AP/AR
Loading and preparing results...
DONE (t=1.84s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *keypoints*
DONE (t=5.02s).
Accumulating evaluation results...
DONE (t=0.16s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.020
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.089
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.002
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.018
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.026
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.040
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.147
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.010
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.030
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.054
best_loss 0.001153 at epoch 1
best_acc 0.232520 at epoch 1
best_mAP 0.020019 at epoch 1
`

**Comment by stefanopini** [comment]
Hi!
I ran the code following your setting and I experienced a similar performance drop (Val AP@IoU=0.50:0.95: 0.089 after one training epoch) using the default learning_rate of `1e-3`.

However, released pre-trained weights are obtained reducing the learning rate during the training and the learning rate at the end of the training is `1e-5`.

It is written in the official paper "Deep High-Resolution Representation Learning for Human Pose Estimation":
> We use the Adam optimizer [32]. The learning schedule follows the setting [72]. The base learning rate is set as 1e−3, and is dropped to 1e−4 and 1e−5 at the170th and 200th epochs, respectively. The training process is terminated within 210epochs.

I ran the code using lr=1e-5 and the accuracy before and after one training epoch are similar.
For a completely-correct restart of the training process we should need also Adam and Batch Norm parameters, but unfortunately they are not available.

Here's the log:
```
[...]

Epoch 1 of 2 @ 2019-10-14 15:57:01
Validating: 100%|██████████| 397/397 [01:43<00:00,  2.51it/s]

Val AP/AR
WARNING:root:json-tricks: numpy scalar serialization is experimental and may work differently in future versions
Loading and preparing results...
DONE (t=0.48s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *keypoints*
DONE (t=4.97s).
Accumulating evaluation results...
DONE (t=0.11s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.744
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.935
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.834
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.712
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.797
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.776
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.946
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.851
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.738
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.833
Training: 100%|█████████████████████████████████████████████████████████████████████| 9363/9363 [1:13:08<00:00,  1.41it/s]

Train AP/AR
Loading and preparing results...
DONE (t=8.65s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *keypoints*
DONE (t=84.88s).
Accumulating evaluation results...
DONE (t=1.93s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.079
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.133
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.077
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.102
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.066
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.226
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.321
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.233
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.215
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.241
best_loss 0.000954 at epoch 1
best_acc 0.873959 at epoch 1
best_mAP 0.744010 at epoch 1

Epoch 2 of 2 @ 2019-10-14 17:19:33
Validating: 100%|██████████| 397/397 [02:55<00:00,  1.97it/s]

Val AP/AR
Loading and preparing results...
DONE (t=0.47s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *keypoints*
DONE (t=2.73s).
Accumulating evaluation results...
DONE (t=0.09s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.717
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.926
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.811
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.688
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.766
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.754
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.939
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.834
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.717
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.809

[...]
```

**Comment by toshi2k2** [comment]
Ah, gotcha. Thank you.
